core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala

```diff
@@ -0,0 +1,552 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.scheduler.cluster.mesos
+
+import java.io.File
+import java.text.SimpleDateFormat
+import java.util.concurrent.atomic.AtomicLong
+import java.util.concurrent.locks.ReentrantLock
+import java.util.{Date, List => JList}
+
+import org.apache.mesos.{SchedulerDriver, Scheduler}
+import org.apache.mesos.Protos._
+import org.apache.spark.deploy.master.DriverState
+import org.apache.spark.deploy.master.DriverState.DriverState
+import org.apache.spark.SparkConf
+import org.apache.spark.SparkException
+import org.apache.spark.util.Utils
+
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+import scala.collection.JavaConversions._
+
+import scala.concurrent.duration.Duration
+import org.apache.mesos.Protos.Environment.Variable
+import org.apache.spark.deploy.mesos.MesosDriverDescription
+import org.apache.mesos.Protos.TaskStatus.Reason
+
+private[spark] class DriverSubmission(
+    val submissionId: String,
+    val desc: MesosDriverDescription,
+    val submitDate: Date) extends Serializable {
+
+  def canEqual(other: Any): Boolean = other.isInstanceOf[DriverSubmission]
+
+  override def equals(other: Any): Boolean = other match {
+    case that: DriverSubmission =>
+      (that canEqual this) &&
+        submissionId == that.submissionId
+    case _ => false
+  }
+}
+
+private [spark] case class ClusterTaskState(
+    val submission: DriverSubmission,
+    val taskId: TaskID,
+    val slaveId: SlaveID,
+    var taskState: Option[TaskStatus],
+    var driverState: DriverState,
+    var startDate: Date,
+    val lastRetry: Option[RetryState] = None) extends Serializable {
+
+  def copy(): ClusterTaskState = {
+    ClusterTaskState(submission, taskId, slaveId, taskState, driverState, startDate, lastRetry)
+  }
+}
+
+private[spark] case class SubmitResponse(id: String, success: Boolean, message: String)
+
+private[spark] case class StatusResponse(
+    id: String,
+    success: Boolean,
+    state: String,
+    status: Option[TaskStatus] = None)
+
+private[spark] case class KillResponse(id: String, success: Boolean, message: String)
+
+private[spark] case class ClusterSchedulerState(
+    appId: String,
+    queuedDrivers: Iterable[DriverSubmission],
+    launchedDrivers: Iterable[ClusterTaskState],
+    finishedDrivers: Iterable[ClusterTaskState],
+    retryList: Iterable[RetryState])
+
+private[spark] trait ClusterScheduler {
```

Comment:
making a parent trait of this might be over engineering this a little. I don't believe we will ever have another `ClusterScheduler` that specifically submits a driver with `MesosDriverDescription`.